### PR TITLE
Additional invoice details via 'Create invoice candidates' endpoint.

### PR DESCRIPTION
**Pull request by pvpurcarcosmin**
https://github.com/metasfresh/metasfresh/issues/5887

### CODE REVIEW

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/InvoiceDetailItem.java  1
- de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JSONInvoiceDetailItem.java  1
- de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/CreateInvoiceCandidatesService.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metasfresh/metasfresh/pullRequest?prid=4641610&bid=15549085)

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/CreateInvoiceCandidatesService.java  1
- de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JSONInvoiceDetailItem.java  1
- de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/InvoiceDetailItem.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metasfresh/metasfresh/pullRequest?prid=4641610&bid=15549085)

File: de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/CreateInvoiceCandidatesService.java [hunk 1]
```diff
@@ -156,6 +155,8 @@ public JsonCreateInvoiceCandidatesResponse createInvoiceCandidates(@NonNull fina
 
 			final InvoiceCandidateId candidateId = externallyReferencedCandidateRepository.save(manualCandidateService.createInvoiceCandidate(candidateToSave));
 
+			externallyReferencedCandidateRepository.storeInvoiceDetailItems(candidateId, candidateToSave.getInvoiceDetailItems());
```
Comment (metas-ts) [inline de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/CreateInvoiceCandidatesService.java]: Saving the candidate's invoiceDetailItems should be done within or below `externallyReferencedCandidateRepository.save(...)`
I.e. the complete "tree" of the invoice candidate shall be persisted in that method

File: de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JsonCreateInvoiceCandidatesRequestItem.java [hunk 2]
```diff
@@ -131,6 +132,10 @@
 			value = "optional invoice line description")
 	String lineDescription;
 
+	@ApiModelProperty(position = 190, required = false, //
+			value = "optional invoice detail items")
```
Comment (metas-ts) [inline de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JsonCreateInvoiceCandidatesRequestItem.java]:
```suggestion
			value = "Optional invoice detail items. Will be persisted as `C_Invoice_Detail` records together with the new invoice candidate.")
```

File: de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/ExternallyReferencedCandidateRepository.java [hunk 3]
```diff
@@ -288,4 +292,34 @@ private ExternallyReferencedCandidate forRecord(@NonNull final I_C_Invoice_Candi
...
 	}
+
+	/**
+	 * Persists the given invoice detail items that are not considered empty.
+	 * @see InvoiceDetailItem#isEmpty()
+	 *
+	 * @param invoiceCandidateId	Id of the invoice candidate to which the detail items are referring.
+	 * @param invoiceDetailItems    TO holding additional details about invoice candidate.
+	 */
+	public void storeInvoiceDetailItems(final InvoiceCandidateId invoiceCandidateId, final List<InvoiceDetailItem> invoiceDetailItems)
```
Comment (metas-ts) [inline de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/ExternallyReferencedCandidateRepository.java]: As pritten further up; Eric Evans would love to see this this method be not public, but called from `save(..)`.
(feel free to contact @teosarca or me for details)

File: de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/InvoiceDetailItem.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+		this.label = label;
+		this.description = description;
+	}
+
+	/**
+	 * @return true if both label and description are blank, false otherwise.
+	 */
+	public boolean isEmpty()
+	{
+		return StringUtils.isBlank(label) && StringUtils.isBlank(description);
```
Comment (metas-ts) [inline de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/InvoiceDetailItem.java]: Note: although we do have apache-StringUtils in the classpath, it's more common in metasfresh-code to use something like `de.metas.util.Check.isEmpty(label, true)`. 
(but for me it's ok like this)

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/InvoiceDetailItem.java  1
- de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/CreateInvoiceCandidatesService.java  1
- de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JSONInvoiceDetailItem.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metasfresh/metasfresh/pullRequest?prid=4641610&bid=15549085)

File: de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/ExternallyReferencedCandidate.java [hunk 5]
```diff
@@ -180,6 +185,7 @@ private ExternallyReferencedCandidate(
 		this.taxId = taxId;
 		this.invoiceDocTypeId = invoiceDocTypeId;
 		this.lineDescription = lineDescription;
+		this.invoiceDetailItems = invoiceDetailItems;
```
Comment (teosarca) [inline de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/ExternallyReferencedCandidate.java]: this.invoiceDetailItems = invoiceDetailItems != null ? ImmutableList.copyOf(invoiceDetailItems) : ImmutableList.of();

Comment (metas-ts) [reply de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/ExternallyReferencedCandidate.java]: done

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/ExternallyReferencedCandidate.java  5
- de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/InvoiceDetailItem.java  1
- de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/CreateInvoiceCandidatesService.java  1
- de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JSONInvoiceDetailItem.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metasfresh/metasfresh/pullRequest?prid=4641610&bid=15549085)

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/ExternallyReferencedCandidate.java  5
- de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/InvoiceDetailItem.java  1
- de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/CreateInvoiceCandidatesService.java  1
- de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JSONInvoiceDetailItem.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metasfresh/metasfresh/pullRequest?prid=4641610&bid=15549085)

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/ExternallyReferencedCandidate.java  5
- de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/invoicecandidates/impl/CreateInvoiceCandidatesService.java  1
- de.metas.business.rest-api/src/main/java/de/metas/rest_api/invoicecandidates/request/JSONInvoiceDetailItem.java  1
- de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/externallyreferenced/InvoiceDetailItem.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/metasfresh/metasfresh/pullRequest?prid=4641610&bid=15549085)

**Pull request merged**